### PR TITLE
[dev] added finish func to consumer

### PR DESCRIPTION
**Pull request by jonaphael**
* **Please check if the PR fulfills these requirements**
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

* **What is the current behavior?** (You can also link to an open issue here)

* **What is the new behavior (if this is a feature change)?**

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

* **Is there any issue related to this PR in other repository?** (such as dojot/dojot)

* **Other information**:

### CODE REVIEW

File: lib/kafka/Consumer.js [hunk 1]
```diff
@@ -292,6 +314,47 @@ module.exports = class Consumer {
...
 
+  /**
+   * Deinit the Consumer
+   *
+   * This method unsubscribe the consumer from all topics and
+   * disconnect it from kafka
+   *
+   * If it was not connected it won't do anything.
+   */
+  deinit() {
```
Comment (mprevide) [inline lib/kafka/Consumer.js:325]: deinit? Maybe disconnect? unsubscribeAll?

File: examples/consumer/sample.js [hunk 2]
```diff
@@ -30,19 +39,20 @@ consumer.init().then(() => {
...
     const { value: payload } = data;
     logger.debug(`Payload: ${payload.toString()}`);
   });
+
+
+  // this example runs for 5 minutes after that its deinit the consumer
+  setTimeout(async () => {
+    clearInterval(getStatusInterval);
+    await consumer.deinit();
+  }, 30000); // 300000
```
Comment (mprevide) [inline examples/consumer/sample.js:55]: Remove // 300000

Comment (eduardogmisiuk) [inline examples/consumer/sample.js:55]: Misleading comment or wrong interval? Remove the `// 300000` comment.

File: lib/kafka/Consumer.js [hunk 3]
```diff
@@ -292,6 +314,47 @@ module.exports = class Consumer {
...
+            logger.error(`Error while disconnecting consumer: ${err}`);
+            return reject(err);
+          }
+          logger.debug('Successfully disconnected producer from Kafka.');
+          this.isReady = false;
+          return resolve();
+        });
+      } catch (error) {
+        return reject(error);
+      }
```
Comment (eduardogmisiuk) [inline lib/kafka/Consumer.js:354]: As I understand Promises, when you call reject or resolve, it already returns the function. Are all those returns in those functions really needed?

**Comment by eduardogmisiuk** [comment]
Change the branch from `master` to `development` and correct the title `develpment` to `dev`.

File: lib/kafka/Consumer.js [hunk 4]
```diff
@@ -292,6 +314,47 @@ module.exports = class Consumer {
...
+      logger.debug('... not initialized. Aborting operation.');
+      return Promise.resolve();
+    }
+
+    return new Promise((resolve, reject) => {
+      logger.debug('Unsubscribing from all topics and disconnecting ...');
+
+      try {
+        this.consumer.unsubscribe();
+        const disconnectTimeout = setTimeout(() => {
```
Comment (rascaraficci) [inline lib/kafka/Consumer.js:337]: I think it is better to set the timer before call unsubscribe().

**Review by rascaraficci — CHANGES_REQUESTED** [review]
I left some comments.

File: lib/kafka/Consumer.js (hunk 1 above)
Comment (mprevide) [reply lib/kafka/Consumer.js:325]: finish looks better

Comment (jonaphael) [reply lib/kafka/Consumer.js:325]: changed to finish!

File: lib/kafka/Consumer.js (hunk 3 above)
Comment (eduardogmisiuk) [reply lib/kafka/Consumer.js:354]: Yes, that's true, I didn't know that. For someone with any doubts about this: https://stackoverflow.com/a/32536083

File: lib/kafka/CommitManager.js [hunk 5]
```diff
@@ -39,13 +39,13 @@ module.exports = class CommitManager {
   }
 
   /**
-   * Deinit the commit manager, clear the interval
+   * finish the commit manager, clear the interval
    */
-  deinit() {
+  finish() {
```
Comment (rascaraficci) [inline lib/kafka/CommitManager.js:44]: Why don't to try to commit the messages here?

File: test/unit/kafka/Consumer.test.js [hunk 6]
```diff
@@ -192,6 +192,47 @@ test('Failed initialization', async () => {
   expect(consumer.isReady).toBe(false);
 });
 
+test('Finish - success even consumer not ready', async () => {
```
Comment (eduardogmisiuk) [inline test/unit/kafka/Consumer.test.js:195]: succeed even when the consumer is not ready

File: lib/kafka/Consumer.js [hunk 7]
```diff
@@ -333,12 +332,16 @@ module.exports = class Consumer {
...
 
       try {
-        this.consumer.unsubscribe();
         const disconnectTimeout = setTimeout(() => {
-          logger.error('Unable to disconnect the consume due to timeout.');
+          logger.error('Unable to disconnect the consumer due to timeout.');
           return reject(new Error('disconnection timeout'));
         }, this.config['disconnect.interval.ms']);
 
+        this.consumer.unsubscribe();
```
Comment (rascaraficci) [inline lib/kafka/Consumer.js:340]: Will this unsubscribe all topics of the consumer? I didn't scrutinize the unsubscribe() method, but seems it is necessary some changes.

File: test/unit/kafka/Consumer.test.js [hunk 8]
```diff
@@ -192,6 +192,47 @@ test('Failed initialization', async () => {
...
+  consumer.disconnect = jest.fn();
+
+  await consumer.finish();
+  expect(consumer.unsubscribe).not.toHaveBeenCalled();
+  expect(consumer.disconnect).not.toHaveBeenCalled();
+});
+
+test('Finish - success', (done) => {
+  const consumer = new Consumer();
+  // consumer.isReady = true;
```
Comment (eduardogmisiuk) [inline test/unit/kafka/Consumer.test.js:208]: Remove it.

File: test/unit/kafka/Producer.test.js [hunk 9]
```diff
@@ -66,6 +67,14 @@ describe('Kafka producer', () => {
   });
 
   describe('Producer creation', () => {
+    it('Should create a producer connect and fail produce', () => {
```
Comment (eduardogmisiuk) [inline test/unit/kafka/Producer.test.js:70]: should create a producer connection and fail to produce

File: test/unit/kafka/Consumer.test.js [hunk 10]
```diff
@@ -191,6 +192,47 @@ test('Failed initialization', async () => {
   expect(consumer.isReady).toBe(false);
 });
 
+test('Finish - success even consumer not ready', async () => {
+  const consumer = new Consumer();
+
+  consumer.unsubscribe = jest.fn();
```
Comment (rascaraficci) [inline test/unit/kafka/Consumer.test.js:198]: It is not a good idea to mock unsubscribe() and disconnect(), you should mock the dependencies of the consumer class.

File: test/unit/kafka/Consumer.test.js [hunk 11]
```diff
@@ -191,6 +192,47 @@ test('Failed initialization', async () => {
...
+
+  await consumer.finish();
+  expect(consumer.unsubscribe).not.toHaveBeenCalled();
+  expect(consumer.disconnect).not.toHaveBeenCalled();
+});
+
+test('Finish - success', (done) => {
+  const consumer = new Consumer();
+  // consumer.isReady = true;
+  consumer.unsubscribe = jest.fn();
```
Comment (rascaraficci) [inline test/unit/kafka/Consumer.test.js:208]: The same comment above.

**Review by rascaraficci — COMMENTED** [review]
I left some comments.

File: lib/kafka/Consumer.js (hunk 7 above)
Comment (jonaphael) [reply lib/kafka/Consumer.js:340]: Yes its unsubscribe from all topics of the consumer. 
For more details, you can check here `https://github.com/Blizzard/node-rdkafka#standard-api-1`

**Review by eduardogmisiuk — APPROVED** [review]
LGTM!

File: test/unit/kafka/Consumer.test.js [hunk 12]
```diff
@@ -191,6 +192,44 @@ test('Failed initialization', async () => {
...
+
+test('Finish - success', (done) => {
+  const consumer = new Consumer();
+  consumer.consumer = new KafkaMock.KafkaConsumer();
+
+  consumer.init();
+  consumer.isReady = true;
+
+
+  consumer.commitManager = new CommitManagerMock();
```
Comment (eduardogmisiuk) [inline test/unit/kafka/Consumer.test.js:212]: As I could see, you did not mock this dependency.

**Review by eduardogmisiuk — COMMENTED** [review]
I left one comment.

**Review by rascaraficci — COMMENTED** [review]
I think that the finish() method of the consumer should also clean its internal data structures too. There are some subscriptions lists and queues. One possibility to clean the internal subscriptions would be call the unsubscribe() instead of the consumer.unsubscribe(), but would be necessary to change the  method unsubscribe().

File: test/unit/kafka/Consumer.test.js (hunk 12 above)
Comment (jonaphael) [reply test/unit/kafka/Consumer.test.js:212]: It's been mocked on line 5
`jest.mock('kafka/CommitManager.js');`

Comment (eduardogmisiuk) [reply test/unit/kafka/Consumer.test.js:212]: That's true, didn't see it

**Comment by jonaphael** [comment]
> I think that the finish() method of the consumer should also clean its internal data structures too. There are some subscriptions lists and queues. One possibility to clean the internal subscriptions would be call the unsubscribe() instead of the consumer.unsubscribe(), but would be necessary to change the method unsubscribe().

The disconnect method fires the `disconnected` event, I'll clean up internal data structures in a callback attached to this event, and it will not be necessary to change or implement a new `unsubscribe` function.

**Review by rascaraficci — APPROVED** [review]
It looks good to me!

**Review by eduardogmisiuk — APPROVED** [review]
LGTM!

**Pull request merged**